meteor/client/ui/Prompter/controller/joycon-device.ts

```diff
@@ -0,0 +1,228 @@
+				if (pad.id && typeof pad.id === 'string') {
+					if (pad.id.match('(L)')) {
+						return pad.axes[0] * -1 // in this mode, L is "negative"
+					} else if (pad.id.match('(R)')) {
+						return pad.axes[0] // in this mode, R is "positive"
+					}
+				}
+			}
+		} else if (pad.axes.length === 4) {
+			// L + R mode
+			// get the first one that is moving outside of the deadband, priorotizing the L controller
+			if (Math.abs(pad.axes[1]) > this.deadBand) {
+				return pad.axes[1] * -1 // in this mode, we are "negative" on both sticks....
+			}
+			if (Math.abs(pad.axes[3]) > this.deadBand) {
+				return pad.axes[3] * -1 // in this mode, we are "negative" on both sticks....
+			}
+		}
+
+		const newButtons = pad.buttons.map(i => i.value)
+
+		if (this.lastButtonArray.length) {
+			for (let i in newButtons) {
+				const oldBtn = this.lastButtonArray[i]
+				const newBtn = newButtons[i]
+				if (oldBtn === newBtn) continue
+
+				// if () { // press
+
+				// } else if () { // release 
+
+				// }
+			}
+		}
+		console.log('')
+
+		this.lastButtonArray = newButtons
+
+		return 0
+	}
+
+	private getSpeedFromJoycons() {
+		const { rangeRevMin, rangeNeutralMin, rangeNeutralMax, rangeFwdMax } = this
+		let inputValue = this.getActiveAxesOfJoycons()
+
+		// start by clamping value to the leagal range
+		inputValue = Math.min(Math.max(inputValue, rangeRevMin), rangeFwdMax) // clamps in between rangeRevMin and rangeFwdMax
+		// stores only for debugging
+		this.lastInputValue = inputValue.toFixed(2)
+
+		if (inputValue >= rangeRevMin && inputValue <= rangeNeutralMin) {
+			// 1) Use the reverse speed spline for the expected speed. The reverse speed is specified using positive values,
+			//    so the result needs to be inversed
+			this.lastSpeed = Math.round(this.reverseSpeedSpline.at(inputValue)) * -1
+		} else if (inputValue >= rangeNeutralMin && inputValue <= rangeNeutralMax) {
+			// 2) we're in the neutral zone
+			this.lastSpeed = 0
+		} else if (inputValue >= rangeNeutralMax && inputValue <= rangeFwdMax) {
+			// 3) Use the speed spline to find the expected speed at this point
+			this.lastSpeed = Math.round(this.speedSpline.at(inputValue))
+		} else {
+			// 4) we should never be able to hit this due to validation above
+			console.error(`Illegal input value ${inputValue}`)
+			return
+		}
+	}
+
+	private updateScrollPosition() {
+		if (this.updateSpeedHandle !== null) return
+		if (!this.getJoycon()) return
+
+		this.getSpeedFromJoycons()
+
+		// update scroll position
+		window.scrollBy(0, this.lastSpeed)
+
+		const scrollPosition = window.scrollY
+		// check for reached end-of-scroll:
+		if (this.currentPosition !== undefined && scrollPosition !== undefined) {
+			if (this.currentPosition === scrollPosition) {
+				// We tried to move, but haven't
+				this.lastSpeed = 0
+
+				// @todo: haptic feedback
+			}
+		}
+		this.currentPosition = scrollPosition
+
+		// debug
+		this.prompterView.DEBUG_controllerState({
```

Comment:
Doing this on every frame might impose some penalty. The `DEBUG_controllerState` is supposed to be very fast, but still, I think it may be best if we somehow limit this to only when `lastSpeed` or `lastInputValue` has changed?